### PR TITLE
Typescript Conversion: main classes

### PR DESCRIPTION
**Pull request by brisberg**
This is a reasonable start, as the tests still pass and we get some type safety on some things.

Unfortunately I have punted some of the work and left in a large number of `any` types in here to get them to compile. Especially around `@screeps/driver`, `@screeps/common`, `@screeps/storage`, and any of the results from a `db` lookup.

I don't think this project is the right place to put bespoke typings for those main projects, but it will take some time to successfully type them. Maybe mongodb typings would help on the database?

--

As discussed in #7, it might make more sense to target this PR into a `typescript` branch, so others can contribute to the typings.

Also we need to convert the tests to TS as well, but that should be straight forward.

All of the options in the `tsconfig.json` are up in the air. I don't feel strongly about any of them, though this configuration does compile and test /shrug.

### CODE REVIEW

File: src/user.ts [hunk 1]
```diff
@@ -1,13 +1,18 @@
...
+    private _id: string;
+    private _username: string;
+    private _server: any;
 
-class User extends EventEmitter {
     /**
         Constructor
     */
-    constructor(server, data) {
+    constructor(server: any, data: any) {
```
Comment (pyrodogg) [inline src/user.ts]: ```server: ScreepsServer```

File: src/world.ts [hunk 2]
```diff
@@ -1,16 +1,28 @@
...
+}
 
-class World {
+export default class World {
+    private server: any;
     /**
         Constructor
     */
-    constructor(server) {
+    constructor(server: any) {
```
Comment (pyrodogg) [inline src/world.ts]: ```server: ScreepsServer```

File: tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,25 @@
+{
+  "compilerOptions": {
+    "module": "CommonJS",
+    "lib": [
+      "esnext"
+    ],
+    "target": "ES6",
```
Comment (pyrodogg) [inline tsconfig.json]: Change target to be at least ES2017, or higher. With ES6, it's polyfilling Promises which isn't necessary.
```suggestion
    "target": "ES2017",
```

Comment (brisberg) [reply tsconfig.json]: Seems to compile/test with ESNext, does that sound good?

Comment (brisberg) [reply tsconfig.json]: Seems to compile/test with ESNext. Does that sound good?

**Comment by brisberg** [comment]
Feel free to merge if `ESNext` is an acceptable target.

**Comment by brisberg** [comment]
Actually, one last thing to review. I removed the `source` flag and added the `declaration` flag to the tsconfig options. This way the compiler will emit a `.d.ts` typings file for the library. Would you prefer to keep the source maps? I don't think they will be needed by consumers on NPM.

**Pull request merged**